src/ocaml_lsp.ml

```diff
@@ -1,65 +1,105 @@
 open! Import
 
+module Experimental_capabilities = struct
+  type t =
+    { interfaceSpecificLangId : bool
+    ; handleSwitchImplIntf : bool
+    ; handleInferIntf : bool
+    ; handleTypedHoles : bool
+    }
+
+  let default =
+    { interfaceSpecificLangId = false
+    ; handleSwitchImplIntf = false
+    ; handleInferIntf = false
+    ; handleTypedHoles = false
+    }
+
+  (** Creates [t] given a JSON of form [{ 'handleSwitchImplIntf' : true, .... }] *)
+  let t_of_json (json : Jsonoo.t) =
+    let experimental_caps_tbl = Jsonoo.Decode.(dict bool) json in
+    let has_capability cap =
+      Stdlib.Hashtbl.find_opt experimental_caps_tbl cap
+      |> Option.value ~default:false
+    in
+    try
+      let interfaceSpecificLangId = has_capability "interfaceSpecificLangId" in
+      let handleSwitchImplIntf = has_capability "handleSwitchImplIntf" in
+      let handleInferIntf = has_capability "handleInferIntf" in
+      let handleTypedHoles = has_capability "handleTypedHoles" in
+      { interfaceSpecificLangId
+      ; handleSwitchImplIntf
+      ; handleInferIntf
+      ; handleTypedHoles
+      }
+    with
+    | Jsonoo.Decode_error err ->
+      show_message `Warn
+        "Unexpected experimental capabilities from the language server. \
+         Falling back to default: no experimental capabilities set.";
+      log_chan `Error ~section:"Ocaml_lsp.of_json" "%s" err;
+      default
+
+  let of_initialize_result (t : LanguageClient.InitializeResult.t) =
+    LanguageClient.InitializeResult.capabilities t
+    |> LanguageClient.ServerCapabilities.experimental
+    |> Option.bind ~f:Jsonoo.Decode.(try_optional (field "ocamllsp" t_of_json))
+    |> Option.value ~default
+end
+
 type t =
-  { interfaceSpecificLangId : bool
-  ; handleSwitchImplIntf : bool
-  ; handleInferIntf : bool
-  ; handleTypedHoles : bool
+  { serverInfo : LanguageClient.InitializeResult.serverInfo option
+  ; experimental_capabilities : Experimental_capabilities.t
   }
 
-let default =
-  { interfaceSpecificLangId = false
-  ; handleSwitchImplIntf = false
-  ; handleInferIntf = false
-  ; handleTypedHoles = false
-  }
+let get_version_from_serverInfo { serverInfo; experimental_capabilities = _ } =
+  match serverInfo with
+  | None -> Error `Missing_serverInfo (* ocamllsp should have [serverInfo] *)
+  | Some { name; version } -> (
+    if not (String.equal name "ocamllsp") then (
+      log_chan ~section:"Ocaml_lsp.get_version" `Warn
+        "the language server is not ocamllsp";
+      (* practically impossible but let's be defensive *)
+      Error `Language_server_isn't_ocamllsp
+    ) else
+      match version with
+      | None -> Error `ServerInfo_version_missing
+      | Some v -> Ok v)
 
-let default_field key decode default json =
-  let open Jsonoo.Decode in
-  try_default default (field key decode) json
+let get_version_semver t =
+  match get_version_from_serverInfo t with
+  | Ok v -> (
+    match String.split v ~on:'-' |> List.hd with
```

Comment:
lsplit2

Reply:
I'm afraid:
`String.lsplit2 "1.9.0" ~on:'-'` evaluates to `None` (because `~on` doesn't occur there), but we want 
`String.split "1.9.0" ~on:'-' |> List.hd` which evaluates to `Some "1.9.0`